pkg/prom/wal/wal.go

```diff
@@ -566,78 +565,66 @@ type appender struct {
-}
+	series := a.w.series.getByID(ref)
+	if series == nil {
+		// Ensure no empty or duplicate labels have gotten through. This mirrors the
+		// equivalent validation code in the TSDB's headAppender.
+		l = l.WithoutEmpty()
+		if len(l) == 0 {
+			return 0, errors.Wrap(tsdb.ErrInvalidSample, "empty labelset")
+		}
 
-func (a *appender) Add(l labels.Labels, t int64, v float64) (uint64, error) {
-	hash := l.Hash()
-	series := a.w.series.getByHash(hash, l)
-	if series != nil {
-		return series.ref, a.AddFast(series.ref, t, v)
-	}
+		if lbl, dup := l.HasDuplicateLabelNames(); dup {
+			return 0, errors.Wrap(tsdb.ErrInvalidSample, fmt.Sprintf(`label name "%s" is not unique`, lbl))
+		}
 
-	// Ensure no empty or duplicate labels have gotten through. This mirrors the
-	// equivalent validation code in the TSDB's headAppender.
-	l = l.WithoutEmpty()
-	if len(l) == 0 {
-		return 0, errors.Wrap(tsdb.ErrInvalidSample, "empty labelset")
-	}
+		var created bool
+		series, created = a.getOrCreate(l)
+		if created {
+			a.series = append(a.series, record.RefSeries{
+				Ref:    series.ref,
+				Labels: l,
+			})
 
-	if lbl, dup := l.HasDuplicateLabelNames(); dup {
-		return 0, errors.Wrap(tsdb.ErrInvalidSample, fmt.Sprintf(`label name "%s" is not unique`, lbl))
+			a.w.metrics.numActiveSeries.Inc()
+			a.w.metrics.totalCreatedSeries.Inc()
+		}
 	}
 
-	a.w.mtx.Lock()
-	ref := a.w.nextRef
-	a.w.nextRef++
-	a.w.mtx.Unlock()
-
-	series = &memSeries{ref: ref, lset: l}
-	series.updateTs(t)
-
-	a.series = append(a.series, record.RefSeries{
-		Ref:    ref,
-		Labels: l,
-	})
-
-	a.w.series.set(hash, series)
-
-	a.w.metrics.numActiveSeries.Inc()
-	a.w.metrics.totalCreatedSeries.Inc()
-	a.w.metrics.totalAppendedSamples.Inc()
-
-	return series.ref, a.AddFast(series.ref, t, v)
-}
-
-func (a *appender) AddFast(ref uint64, t int64, v float64) error {
-	series := a.w.series.getByID(ref)
-	if series == nil {
-		return storage.ErrNotFound
-	}
 	series.Lock()
 	defer series.Unlock()
 
 	// Update last recorded timestamp. Used by Storage.gc to determine if a
-	// series is dead.
+	// series is stale.
 	series.updateTs(t)
 
 	a.samples = append(a.samples, record.RefSample{
-		Ref: ref,
+		Ref: series.ref,
 		T:   t,
 		V:   v,
 	})
 
 	a.w.metrics.totalAppendedSamples.Inc()
-	return nil
+	return series.ref, nil
+}
+
+func (a *appender) getOrCreate(l labels.Labels) (series *memSeries, created bool) {
+	hash := l.Hash()
+
+	series = a.w.series.getByHash(hash, l)
+	if series != nil {
+		return series, false
+	}
+
+	series = &memSeries{ref: a.w.ref.Inc(), lset: l}
+	a.w.series.set(l.Hash(), series)
+	return series, true
 }
 
 func (a *appender) AppendExemplar(ref uint64, _ labels.Labels, e exemplar.Exemplar) (uint64, error) {
 	s := a.w.series.getByID(ref)
 	if s == nil {
-		return 0, fmt.Errorf("unknown series ref. when trying to add exemplar: %d", ref)
+		return 0, errors.Wrapf(storage.ErrNotFound, "unknown series ref %d when trying to add exemplar", ref)
```

Comment:
Oh, I see, this is taken right out of the TSDB Head. I'll revert this line.

```suggestion
		return 0, fmt.Errorf("unknown series ref. when trying to add exemplar: %d", ref)
```